objects/src/accounts/code/mod.rs

```diff
@@ -42,36 +42,76 @@ impl AccountCode {
-    ///   offset of faucet procedures. This is needed because of the faucet reserved storage slot at
-    ///   location 0.
+    /// This does not check whether the provided components are valid when combined. Prefer
+    /// [`Account::initialize_from_components`](crate::accounts::Account::initialize_from_components)
+    /// whenever possible as it implements that check.
     ///
     /// # Errors
-    /// - If the number of procedures exported from the provided library is 0.
-    /// - If the number of procedures exported from the provided library is greater than 256.
-    /// - If the creation of a new `AccountProcedureInfo` fails.
-    pub fn new(library: Library, is_faucet: bool) -> Result<Self, AccountError> {
-        // extract procedure information from the library exports
-        // TODO: currently, offsets for all regular account procedures are set to 0 and offsets for
-        // faucet accounts procedures are set to 1. Furthermore sizes are set to 1 for all
-        // procedures. Instead they should be read from the Library metadata.
+    ///
+    /// Returns an error if:
+    /// - The number of procedures in all merged libraries is 0 or exceeds
+    ///   [`AccountCode::MAX_NUM_PROCEDURES`].
+    /// - If two or more libraries export a procedure with the same MAST root.
+    /// - If the number of [`StorageSlot`](crate::accounts::StorageSlot)s of a component or of all
+    ///   components exceeds 255.
+    /// - If [`MastForest::merge`] fails on all libraries.
+    pub fn from_components(
+        components: &[AccountComponent],
+        account_type: AccountType,
+    ) -> Result<Self, AccountError> {
+        let (merged_mast_forest, _) =
+            MastForest::merge(components.iter().map(|component| component.mast_forest()))
+                .map_err(|err| AccountError::AccountCodeMergeError(err.to_string()))?;
+
         let mut procedures = Vec::new();
-        let storage_offset = if is_faucet { 1 } else { 0 };
-        let storage_size = 1;
-        for module in library.module_infos() {
-            for proc_mast_root in module.procedure_digests() {
-                procedures.push(AccountProcedureInfo::new(
-                    proc_mast_root,
-                    storage_offset,
-                    storage_size,
-                )?);
+        let mut proc_root_set = BTreeSet::new();
+
+        // Slot 0 is globally reserved for faucet accounts so the accessible slots begin at 1 if
+        // there is a faucet component present.
+        let mut component_storage_offset = if account_type.is_faucet() { 1 } else { 0 };
+
+        for component in components {
+            let AccountComponent { library, storage_slots, .. } = component;
+            let component_storage_size: u8 = storage_slots
+                .len()
+                .try_into()
+                .map_err(|_| AccountError::StorageTooManySlots(storage_slots.len() as u64))?;
```

Comment:
As I mentioned in the PR description, there are other checks regarding the number of procedures in a component's library that could be added in `AccountComponent::new` as well.
However making the constructor fallible means one of two things for `AccountBuilder` as a consequence (which may not be bad, just pointing them out):
- Any implementation of `From<MyComponent> for AccountComponent` has to `expect` the call to `AccountComponent::new` to assert that it is a valid component. That might be fine as the component should be able to assert that its number of storage slots fits in a u8 and potentially others (like `0 < num_procedures < 256`).
- We change the pattern to use `TryFrom` instead. Probably unnecessary as that just propagates errors that should be assertable statically, at least for now. Depending on whether we define `add_component` to take an `AccountComponent` or to take `impl TryInto<AccountComponent>` it would look different:
```rust
AccountBuilder::new().add_component(MyComponent::new(...).try_into()?).build();
AccountBuilder::new().add_component(MyComponent::new(...))?.build();
```
But I don't like both of these. We would just be propagating errors that should have been dealt with earlier. I think the first bullet point is preferable. Most custom components that can actually fail to convert to a valid account component should catch that in _their_ constructor, e.g. `MyComponent::new`, so it would simply be:

```rust
AccountBuilder::new().add_component(MyComponent::new(...)?).build();
```

Which is all to say that: We can add the validation as you suggested in `AccountComponent::new` - I just wanted to explore this and convince myself that it's fine :sweat_smile:.